pkg/workspace/controllers/noderesource/node_resource_controller.go

```diff
@@ -0,0 +1,411 @@
+		readyNodeNames, needsWorkerNodeUpdate, needsResourceStatusUpdate)
+}
+
+// updateWorkspaceStatusWithBothUpdates updates both WorkerNodes and ResourceStatus condition in a single request
+func (c *NodeResourceReconciler) updateWorkspaceStatusWithBothUpdates(ctx context.Context, name *client.ObjectKey,
+	workerNodes []string, updateWorkerNodes bool, updateResourceStatus bool) error {
+	return retry.OnError(retry.DefaultRetry,
+		func(err error) bool {
+			return apierrors.IsServiceUnavailable(err) || apierrors.IsServerTimeout(err) || apierrors.IsTooManyRequests(err) || apierrors.IsConflict(err)
+		},
+		func() error {
+			// Read the latest version to avoid update conflict.
+			wObj := &kaitov1beta1.Workspace{}
+			if err := c.Client.Get(ctx, *name, wObj); err != nil {
+				if !apierrors.IsNotFound(err) {
+					return err
+				}
+				return nil
+			}
+
+			// Update WorkerNodes if needed
+			if updateWorkerNodes {
+				wObj.Status.WorkerNodes = workerNodes
+			}
+
+			// Update ResourceStatus condition if needed
+			if updateResourceStatus {
+				condition := metav1.Condition{
+					Type:               string(kaitov1beta1.ConditionTypeResourceStatus),
+					Status:             metav1.ConditionTrue,
+					Reason:             "ResourcesReady",
+					ObservedGeneration: wObj.GetGeneration(),
+					Message:            "All resources are ready and nodes are available",
+					LastTransitionTime: metav1.Now(),
+				}
+				meta.SetStatusCondition(&wObj.Status.Conditions, condition)
+			}
+
+			return c.Client.Status().Update(ctx, wObj)
+		})
+}
+
+// resetResourceStatusConditionIfNeeded resets the ResourceStatus condition to false if it's currently true
+// when NodeClaim condition is not ready
+func (c *NodeResourceReconciler) resetResourceStatusConditionIfNeeded(ctx context.Context, wObj *kaitov1beta1.Workspace) error {
+	// Check if ResourceStatus condition is currently true
+	resourceCondition := meta.FindStatusCondition(wObj.Status.Conditions, string(kaitov1beta1.ConditionTypeResourceStatus))
+	if resourceCondition == nil || resourceCondition.Status != metav1.ConditionTrue {
+		// Already false or doesn't exist, no need to update
+		return nil
+	}
+
+	// Update ResourceStatus condition to false since NodeClaim condition is not ready
+	klog.InfoS("Resetting ResourceStatus condition to false due to NodeClaim not ready",
+		"workspace", klog.KObj(wObj))
+
+	return ctrlutils.UpdateStatusConditionIfNotMatch(ctx, c.Client, wObj, kaitov1beta1.ConditionTypeResourceStatus, metav1.ConditionFalse,
+		"NodeClaimNotReady", "Waiting for NodeClaims to be ready before resources can be provisioned")
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (c *NodeResourceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	c.Client = mgr.GetClient()
+
+	builder := ctrl.NewControllerManagedBy(mgr).
+		Named("NodeResourceController").
+		For(&kaitov1beta1.Workspace{}).
```

Comment:
What is the reason for having 3 separate controllers reconciling for the same Workspace? Isn't there going to be a lot of stale API server cache / update conflict issues?

Reply:
The best practice of controller is: each controller should focus a specific feature, and the current status for workspace controller is complicated, because it includes nodeclaims management, node addons ensuring, tuning job, inference workloads, etc.

Reply:
It will not cause stale metadata cache, because all metdata cache is managed by controller-runtime, and each resource has only one copy of cache.

Reply:
@chewong My thought is not correct. Yes, Multiple controllers use the same resource(workspace) to coordinate workflow will casue race conflict. I'll go to change these controllers(nodeclaim, noderesource) into libraries.